### PR TITLE
[Home/Artist] Add follow action to artists in artist rail.

### PR DESCRIPTION
**Pull request by alloy**
follow artist

* The Relay warnings shown are because I’m not using Relay to populate the components with data. You can ignore them.
* The layout breaks with some artists, not sure if that’s to do with the replacing or just the data they have. Needs to be inspected further.

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">The following files do not have tests: <a href="https://github.com/artsy/emission/blob/b5f9255584d1a4030ebd03adab0e14fd12a1e634/lib/components/artwork_grids/generic_grid.js">lib/components/artwork_grids/generic_grid.js</a>, <a href="https://github.com/artsy/emission/blob/b5f9255584d1a4030ebd03adab0e14fd12a1e634/lib/components/artwork_grids/infinite_scroll_grid.js">lib/components/artwork_grids/infinite_scroll_grid.js</a>, <a href="https://github.com/artsy/emission/blob/b5f9255584d1a4030ebd03adab0e14fd12a1e634/lib/components/home/artist_rails/artist_card.js">lib/components/home/artist_rails/artist_card.js</a>, <a href="https://github.com/artsy/emission/blob/b5f9255584d1a4030ebd03adab0e14fd12a1e634/lib/components/home/artist_rails/artist_rail.js">lib/components/home/artist_rails/artist_rail.js</a>, <a href="https://github.com/artsy/emission/blob/b5f9255584d1a4030ebd03adab0e14fd12a1e634/lib/components/spinner.js">lib/components/spinner.js</a> &amp; <a href="https://github.com/artsy/emission/blob/b5f9255584d1a4030ebd03adab0e14fd12a1e634/lib/metaphysics.js">lib/metaphysics.js</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: lib/components/home/artist_rails/artist_rail.js [hunk 1]
```diff
@@ -12,17 +12,42 @@ import ArtistCard from './artist_card'
...
+    }
+  }
 
   componentDidMount() {
     this.props.relay.setVariables({ fetchContent: true })
   }
 
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.rail.results) {
+      this.setState({ artists: nextProps.rail.results.slice(0) })
```
Comment (sarahscott) [inline lib/components/home/artist_rails/artist_rail.js]: What is `slice(0)` for in this case?

File: Pod/Classes/SpinnerComponent/ARSpinnerManager.m [hunk 2]
```diff
@@ -8,6 +8,11 @@ @implementation ARSpinnerManager
 
 RCT_EXPORT_MODULE();
 
+RCT_CUSTOM_VIEW_PROPERTY(spinnerColor, NSNumber, ARSpinner)
+{
+  view.spinnerColor = [RCTConvert UIColor:json];
```
Comment (sarahscott) [inline Pod/Classes/SpinnerComponent/ARSpinnerManager.m]: 👏 This will be useful.

File: lib/components/artwork_grids/infinite_scroll_grid.js [hunk 3]
```diff
@@ -67,21 +68,8 @@ class InfiniteScrollArtworksGrid extends React.Component {
...
-        } else {
-          const error = new Error(response.statusText)
-          error.response = response
-          throw error
-        }
-      })
-      .then((response) => response.json())
-      .then(({ data }) => {
-        const artworks = data.artist.artworks
+    metaphysics(query)
```
Comment (sarahscott) [inline lib/components/artwork_grids/infinite_scroll_grid.js]: This looks way nicer.

File: lib/components/home/artist_rails/artist_card.js [hunk 4]
```diff
@@ -33,6 +33,7 @@ class ArtistCard extends React.Component {
     ARTemporaryAPIModule.setFollowStatus(true, this.props.artist._id, (error, following) => {
       if (error) {
         __DEV__ ? console.error(error) : console.log(error)
+        this.setState({ processingChange: false })
```
Comment (sarahscott) [inline lib/components/home/artist_rails/artist_card.js]: Would it make sense to set `processingChange: true` before the networking request? I know these are completion blocks and `setState` will happen right away but conceptually doesn't the change-processing start as soon as the button is pressed?

File: lib/components/home/artist_rails/artist_rail.js (hunk 1 above)
Comment (alloy) [inline lib/components/home/artist_rails/artist_rail.js]: It’s a JS convention to create a copy of an array. Normally you would use `slice` to get a subarray starting from an element at a certain index, so using 0 means you get all of the elements in a new array.

File: lib/components/home/artist_rails/artist_card.js (hunk 4 above)
Comment (alloy) [inline lib/components/home/artist_rails/artist_card.js]: So you mean to move the last line in this method to be before the networking request, right, not this line that’s shown in the review? I think that makes sense, will do.

File: lib/components/home/artist_rails/artist_rail.js (hunk 1 above)
Comment (alloy) [inline lib/components/home/artist_rails/artist_rail.js]: It is actually not really needed in this location, I do it below in `handleFollowChange` again before making a change to the array. I’ll get rid of it here.

Comment (sarahscott) [inline lib/components/home/artist_rails/artist_rail.js]: Good to know!

**Comment by sarahscott** [comment]
👏 👏 👏

**Pull request merged**